### PR TITLE
Changes to logging for search query (SELECT)

### PR DESCRIPTION
**Pull request by patsonluk**
[CH160298]

## Description
The existing logging behaviors for search query:
1. The logging for query hitting solr the first time (for example hitting QA node) would NOT have full query details logged on completion. 
2. The logging for shard query distributed by QA node would have full query details logged on completion. Within the reported `params`, there could be duplication of information in `fq` and `json`.

While the ideal behaviors are:
1. The logging for query hitting solr the first time SHOULD have full query details logged **before the query handling**. This is useful if any query handling stalls and never completes
2. The logging for shard query is too verbose as the full query is printed for each shard query, which could have hundreds of shard for a single query. Ideally we don't want to log `fq` and `json`.

## Cause
1. Missing full query details on first query is due to the fact that `rsp.toLog` (which is a named list of KVs to be logged) is [populated](https://github.com/fullstorydev/lucene-solr/blob/086cb3454caa06f98c28ee3824bb6f3f124fac59/solr/core/src/java/org/apache/solr/core/SolrCore.java#L2704)  BEFORE the actual query is [constructed](https://github.com/fullstorydev/lucene-solr/blob/086cb3454caa06f98c28ee3824bb6f3f124fac59/solr/core/src/java/org/apache/solr/request/json/RequestUtil.java#L71). Therefore on completion, the `params` in `toLog` entry would NOT have the query in there
2. When the `SearchHandler` first handles the request at QA node, it [adds `fq`/`json` as param](https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/request/json/RequestUtil.java#L211), it then distributes the requests along with those params to other shards. Other shards simply print out everything from params, hence the full query in `fq` and `json`

## Solution
1. Added a new `INFO` logging on `SearchHandler#handleRequestBody` with the full query and request id (`rid`). Take note that this will only be executed once when the query first reaches solr (not a re-distributed one). This is done by checking if `distrib` is absent in params (that means it's a plain/direct query to shard, not distributed from other nodes) OR if `distrib` is true, that means this req should be re-distributed (for example from QA node) to other shards
2. If a req is identified as a re-distributed one from QA node, we will replace the verbose `params` in `rsp.toLog` with a compact version with `fq` and `json` values removed from the `params` 

### Before this change
- On SELECT query start (QA node) : **nothing is printed**
- On SELECT query end (QA node):
`
2021-08-31 22:12:00.207 INFO  (qtp1238616099-19) [c:local  r:127.0.0.1 x:local ] o.a.s.c.S.Request [local]  webapp=/solr path=/select params={indent=off&shards.tolerant=true&wt=json} rid=127.0.0.1-0 rid=127.0.0.1-0 status=0 QTime=1303
`
- On SELECT query end (each shard):
`
2021-08-31 22:12:00.167 INFO  (qtp1238616099-295) [c:local s:shard1_0 r:core_node5 x:local_shard1_0_replica_n3 ] o.a.s.c.S.Request [local_shard1_0_replica_n3]  webapp=/solr path=/select params={group.distributed.first=true&distrib=false&TZ=America/Vancouver&fl=Id,score&shards.purpose=2048&fq={!multijoin+to%3D"IndvId"+from%3D"IndvId"+qs%3D"(Kind:indv+AND+UserEmailStr:patson@fullstory\\.com)"}&fq=Kind:event&shard.url=http://127.0.0.1:8983/solr/local_shard1_0_replica_n3/&group.ngroups=true&rid=127.0.0.1-0&group.sort=&shards.tolerant=true&json={%0a++"query":+"Kind:event",%0a++"filter":+[%0a++++"{!multijoin+to%3D\"IndvId\"+from%3D\"IndvId\"+qs%3D\"(Kind:indv+AND+UserEmailStr:patson@fullstory\\\\.com)\"}",%0a++++"Kind:event"%0a++],%0a++"limit":+0,%0a++"facet":+{%0a++++"_f0":+"unique(UserIdSessionId)"%0a++},%0a++"params":+{%0a++++"NOW":+"1630447897016",%0a++++"TZ":+"America/Vancouver",%0a++++"echoParams":+"none",%0a++++"group":+true,%0a++++"group.field":+[%0a++++++"IndvId"%0a++++],%0a++++"group.limit":+0,%0a++++"group.main":+false,%0a++++"group.ngroups":+true,%0a++++"group.sort":+"",%0a++++"wat":+"TimeRangeCounts_MatchingCounts"%0a++}%0a}&wt=javabin&group.field=IndvId&group=true&group.limit=0&start=0&wat=TimeRangeCounts_MatchingCounts&rows=0&version=2&q=Kind:event&group.main=false&omitHeader=false&NOW=1630447897016&isShard=true} rid=127.0.0.1-0 status=0 QTime=2
`

### After this change
- On SELECT query start (QA node) :
`
2021-09-01 00:46:47.475 INFO  (qtp1238616099-19) [c:local  r:127.0.0.1 x:local ] o.a.s.h.c.SearchHandler Start Search Query: indent=off&shards.tolerant=true&json={%0a++"query":+"Kind:event",%0a++"filter":+[%0a++++"{!multijoin+to%3D\"IndvId\"+from%3D\"IndvId\"+qs%3D\"(Kind:indv+AND+UserEmailStr:patson@fullstory\\\\.com)\"}",%0a++++"(Kind:event+AND+EventStart:[NOW\\/DAY\\-1DAY+TO+NOW\\/DAY})",%0a++++"{!terms+f%3DIndvId}7"%0a++],%0a++"limit":+0,%0a++"facet":+{%0a++++"_f0":+{%0a++++++"type":+"terms",%0a++++++"field":+"IndvId",%0a++++++"mincount":+1,%0a++++++"limit":+1,%0a++++++"facet":+{%0a++++++++"_f0":+"unique(UserIdSessionId)"%0a++++++}%0a++++}%0a++},%0a++"params":+{%0a++++"NOW":+"1630457197016",%0a++++"TZ":+"America/Vancouver",%0a++++"echoParams":+"none",%0a++++"wat":+"fetchSessionCountsForIndvs"%0a++}%0a}&wt=json rid=127.0.0.1-9
`
- On SELECT query end (QA node): **(no change)**
`
2021-09-01 00:46:47.485 INFO  (qtp1238616099-19) [c:local  r:127.0.0.1 x:local ] o.a.s.c.S.Request [local]  webapp=/solr path=/select params={indent=off&shards.tolerant=true&wt=json} rid=127.0.0.1-9 rid=127.0.0.1-9 hits=374 status=0 QTime=11
`
- On SELECT query end (shard):
`
021-09-01 00:46:47.483 INFO  (qtp1238616099-86) [c:local s:shard1_1 r:core_node6 x:local_shard1_1_replica_n4 ] o.a.s.c.S.Request [local_shard1_1_replica_n4]  webapp=/solr path=/select params={distrib=false&_facet_={}&echoParams=explicit&TZ=America/Vancouver&fl=Id&fl=score&shards.purpose=1064964&start=0&fsv=true&wat=fetchSessionCountsForIndvs&shard.url=http://127.0.0.1:8983/solr/local_shard1_1_replica_n4/&rows=0&rid=127.0.0.1-9&version=2&q=Kind:event&shards.tolerant=true&omitHeader=false&NOW=1630457197016&isShard=true&wt=javabin} rid=127.0.0.1-9 hits=374 status=0 QTime=4
`

### CODE REVIEW

**Comment by patsonluk** [comment]
@noblepaul any thoughts about this proposed change? Many thanks!

File: solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java [hunk 1]
```diff
@@ -284,6 +286,24 @@ protected ResponseBuilder newResponseBuilder(SolrQueryRequest req, SolrQueryResp
   @SuppressWarnings({"unchecked", "rawtypes"})
   public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throws Exception
   {
+    final boolean ridTaggingDisabled = req.getParams().getBool(CommonParams.DISABLE_REQUEST_ID, false);
+    final String reqId = ridTaggingDisabled ? null : getOrGenerateRequestId(req);
+
+    Boolean distribOjbect = req.getParams().getBool(DISTRIB);
+    if (distribOjbect == null || distribOjbect) {
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:293]: `if (solrReq.getParams().getBool(DISTRIB, false)`
is enough?

File: solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java [hunk 2]
```diff
@@ -530,18 +552,36 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
     }
   }
 
-  private void tagRequestWithRequestId(ResponseBuilder rb) {
-    final boolean ridTaggingDisabled = rb.req.getParams().getBool(CommonParams.DISABLE_REQUEST_ID, false);
-    if (! ridTaggingDisabled) {
-      String rid = getOrGenerateRequestId(rb.req);
-      if (StringUtils.isBlank(rb.req.getParams().get(CommonParams.REQUEST_ID))) {
-        ModifiableSolrParams params = new ModifiableSolrParams(rb.req.getParams());
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:538]: We are making a copy of all the params. If possible we should just avoid it. let's make a new impl of SolrParams which just wraps the existing?

File: solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java [hunk 3]
```diff
@@ -530,18 +552,36 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
...
+      public String[] getParams(String param) {
+        return params.getParams(param);
+      }
+    };
+    return filteredParams;
+  }
+
+  private void tagRequestWithRequestId(ResponseBuilder rb, String rid) {
+    if (StringUtils.isBlank(rb.req.getParams().get(CommonParams.REQUEST_ID))) {
+      ModifiableSolrParams params = new ModifiableSolrParams(rb.req.getParams());
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:579]: Here too let's try to avoid creation of a new Map

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:579]: Thanks noble! What I did was removing the `if (! ridTaggingDisabled) {` check in `tagRequestWithRequestId`, the new Map creation was in the old code and was left as-is.

I am open to improve existing code, but I'm pretty new to the team so I don't know the preference on changes. Should we only do changes related to the goals outlined in this PR or is it okay to do other changes along the way? Perhaps @hiteshk25 can shed some light? 😊

File: solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java (hunk 1 above)
Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:293]: I am thinking there could be cases if someone query the shard directly w/o providing any `DISTRIB` param at all? In such case, I assume we still want full query. Is that a valid use case? Sorry, I don't have enough Solr experience to tell 😓

**Comment by patsonluk** [comment]
@claygoddard We are wondering if your team has any concerns, as with this proposed change the logging on Solrcloud/QA nodes will change. Please read the first comment for change summary and the example of how they will look like after the change 

Please lemme know if you have any questions/concerns! 😊

File: solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java (hunk 1 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:293]: if it is absent, the default is returned

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:293]: Ah ic :) so in such a case we should use `if (solrReq.getParams().getBool(DISTRIB, true))` ? If there are any query hitting the node directly without even specifying the `DISTRIB` param, we should still log full query?

**Comment by claygoddard** [comment]
> @claygoddard We are wondering if your team has any concerns, as with this proposed change the logging on Solrcloud/QA nodes will change. Please read the first comment for change summary and the example of how they will look like after the change
> 
> Please lemme know if you have any questions/concerns! 😊

Description makes sense to me and seems good 👍. My team rarely to never uses these logs.

**Comment by hiteshk25** [comment]
@patsonluk not sure about this but we should prefix host with rid? that will help us to know the #requests from QA node x .

**Comment by patsonluk** [comment]
> @patsonluk not sure about this but we should prefix host with rid? that will help us to know the #requests from QA node x .

Thanks @hiteshk25 . I'm going to close this as very similar change is put into https://github.com/fullstorydev/lucene-solr/pull/102/files#diff-d6fc591739d3396fb74758c164b4631785e059923eb370c9867771ccdf639675R296 and merged. Should close this to avoid confusion! 😅 

I think your concern has been briefly discussed in [here](https://github.com/cowpaths/mn/pull/37857)
>Request ID carries extra info that's not currently available in OpenCensus trace ID/Span ID
Request ID is in format of "%s:%d:%d", gcloud.Env().Hostname(), processId, atomic.AddInt64(&requestIdCounter, 1)) , which OC id has no info of hostname nor process ID

There are several options to get that info back:
1. During the OpenCensus ID -> rid conversion, prepend the hostname like before https://github.com/cowpaths/mn/pull/37857/files#diff-7b0507bcfd706a0583f64273a7a417d8cf23b614cdb73f1910288965cdc225dfR23
2. On the solr side, read the http request header for remote host info and either add that to `rid` or log it somewhere

I think option 1 is probably a bit more trivial. Added story https://app.shortcut.com/fullstory/story/177439/add-host-info-to-rid-passed-to-solr-node so we wouldn't forget 😊

**Pull request closed without merging**